### PR TITLE
Fix alltoallv with inplace

### PR DESCRIPTION
**Pull request by brminich**
Fix handling of mpi_inplace in MPI_Alltoallv. Some of the counts (but not all) can be 0 and the rank can exit at `(0 == max_size)` check, which leads to a hang

The failure is found with https://github.com/pmodels/mpich/blob/main/test/mpi/coll/alltoallv.c

### CODE REVIEW

File: ompi/mca/coll/base/coll_base_alltoallv.c [hunk 1]
```diff
@@ -66,8 +66,15 @@ mca_coll_base_alltoallv_intra_basic_inplace(const void *rbuf, const int *rcounts
     size = ompi_comm_size(comm);
     rank = ompi_comm_rank(comm);
 
-    ompi_datatype_type_size(rdtype, &max_size);
-    max_size *= rcounts[rank];
+    for (i = 0, max_size = 0 ; i < size ; ++i) {
+        if (i == rank) {
+            continue;
+        }
+        ompi_datatype_type_size(rdtype, &packed_size);
```
Comment (bosilca) [inline ompi/mca/coll/base/coll_base_alltoallv.c:73]: I know the datatype cache its size, but we can avoid the unnecessary call to ompi_datatype_type_size by calling it once outside the loop

Comment (brminich) [reply ompi/mca/coll/base/coll_base_alltoallv.c]: fixed, but `ompi_datatype_type_size` is supposed to be inlined

File: ompi/mca/coll/base/coll_base_alltoallv.c [hunk 2]
```diff
@@ -58,16 +58,22 @@ mca_coll_base_alltoallv_intra_basic_inplace(const void *rbuf, const int *rcounts
...
+    ompi_datatype_type_size(rdtype, &type_size);
 
-    ompi_datatype_type_size(rdtype, &max_size);
-    max_size *= rcounts[rank];
+    for (i = 0, max_size = 0 ; i < size ; ++i) {
+        if (i == rank) {
+            continue;
+        }
+        packed_size = rcounts[i] * type_size;
+        max_size = packed_size > max_size ? packed_size : max_size;
```
Comment (hoopoepg) [inline ompi/mca/coll/base/coll_base_alltoallv.c:75]: there is no any ```max``` macro?

**Review by hoopoepg — APPROVED** [review]

File: ompi/mca/coll/base/coll_base_alltoallv.c (hunk 2 above)
Comment (ggouaillardet) [reply ompi/mca/coll/base/coll_base_alltoallv.c:75]: yes, it is `opal_max()` in `opal/util/minmax.h`

File: ompi/mca/coll/base/coll_base_alltoallv.c (hunk 1 above)
Comment (bosilca) [reply ompi/mca/coll/base/coll_base_alltoallv.c]: a linearly with rank number of unnecessary calls.

**Review by bosilca — APPROVED** [review]

File: ompi/mca/coll/base/coll_base_alltoallv.c (hunk 2 above)
Comment (brminich) [reply ompi/mca/coll/base/coll_base_alltoallv.c]: replaced with `opal_max`

**Review by hoopoepg — APPROVED** [review]

**Comment by awlauria** [comment]
@brminich can you port to the release branches? v4/4.1/5

**Comment by brminich** [comment]
> @brminich can you port to the release branches? v4/4.1/5

#10073 and #10072

this issue is not present in v4.0.x

**Pull request merged**